File: LeetCode/Hard/1948-Delete_Duplicate_Folders_In_System/1948-DeleteDuplicateFoldersInSystem.py
from collections import defaultdict


class Trie:
    def __init__(self, folder):
        self.folder = folder
        self.remove = False
        self.childs = defaultdict()


class Solution:
    def deleteDuplicateFolder(self, paths: list[list[str]]) -> list[list[str]]:
        def insert(root, path):
            for folder in path:
                if folder not in root.childs:
                    root.childs[folder] = Trie(folder)

                root = root.childs[folder]

        def mark_duplicate(root):
            subfolder = []

            for _, child_node in sorted(root.childs.items()):
                subfolder.append(mark_duplicate(child_node))

            code = "".join(subfolder)

            if subfolder:
                if code in visited:
                    visited[code].remove = True
                    root.remove = True
                else:
                    visited[code] = root

            return f"[{root.folder}{code}]"

        def remove_duplicates(root):
            if root.remove:
                return

            stack.append(root.folder)
            ans.append(stack.copy())

            for child_node in root.childs.values():
                remove_duplicates(child_node)

            stack.pop()

        root = Trie("/")

        # Add all paths to the Trie tree
        for path in paths:
            insert(root, path)

        # Mark the duplicates folders
        visited = {}
        mark_duplicate(root)

        # Remove the duplicate folders
        ans = []
        stack = []
        for child_node in root.childs.values():
            remove_duplicates(child_node)

        return ans


a = Solution()
print(
    a.deleteDuplicateFolder(
        paths=[["a"], ["c"], ["d"], ["a", "b"], ["c", "b"], ["d", "a"]]
    )
)
print(
    a.deleteDuplicateFolder(
        paths=[
            ["a"],
            ["c"],
            ["a", "b"],
            ["c", "b"],
            ["a", "b", "x"],
            ["a", "b", "x", "y"],
            ["w"],
            ["w", "y"],
        ]
    )
)
print(a.deleteDuplicateFolder(paths=[["a", "b"], ["c", "d"], ["c"], ["a"]]))

File: LeetCode/Hard/1948-Delete_Duplicate_Folders_In_System/1948-DeleteDuplicateFoldersInSystem.js
class TrieNode {
    constructor(folder_name) {
        this.folder = folder_name;
        this.remove = false;
        this.childs = new Map();
    }
}

var insert = function (root, path) {
    for (const folder of path) {
        if (!root.childs.has(folder)) {
            root.childs.set(folder, new TrieNode(folder));
        }
        root = root.childs.get(folder);
    }
}

var mark_duplicate = function (root, visited) {
    let subfolder = [];
    const sortedChildKeys = Array.from(root.childs.keys()).sort();
    for (const child_folder of sortedChildKeys) {
        const child_node = root.childs.get(child_folder);
        subfolder.push(mark_duplicate(child_node, visited));
    }

    let code = subfolder.join("");

    if (code.length > 0) {
        if (visited.has(code)) {
            visited.get(code).remove = true;
            root.remove = true;
        } else {
            visited.set(code, root);
        }
    }

    return `[${root.folder}${code}]`;
}

var remove_duplicates = function (root, stack, result) {
    if (root.remove) {
        return;
    }

    if (root.folder !== "/") {
        stack.push(root.folder);
        result.push([...stack]);
    }

    for (const child_folder of root.childs.keys()) {
        const child_node = root.childs.get(child_folder);
        remove_duplicates(child_node, [...stack], result);
    }

    if (root.folder !== "/") {
        stack.pop();
    }
}

/**
 * @param {string[][]} paths
 * @return {string[][]}
 */
var deleteDuplicateFolder = function (paths) {
    // Add all paths to the Trie tree
    const root = new TrieNode("/");
    for (const path of paths) {
        insert(root, path);
    }

    // Mark the duplicates folders
    const visited = new Map();
    mark_duplicate(root, visited);

    // Remove the duplicate folders
    const result = [];
    for (const folder of root.childs.values()) {
        remove_duplicates(folder, [], result);
    }

    return result;
};